subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/TestExecutionBuildConfigurationAction.java

```diff
@@ -112,29 +114,52 @@ private List<Test> configureBuildForTestDescriptors(TestExecutionRequestAction t
             }
         }
 
-        for (InternalTestPatternSpec patternSpec : testExecutionRequest.getTestPatternSpecs()) {
-            for (Test task : queryTestTasks(patternSpec.getTaskPath())) {
-                testTasksToRun.add(task);
-                TestFilter filter = task.getFilter();
-                for (String cls : patternSpec.getClasses()) {
-                    filter.includeTest(cls, null);
-                }
-                for (Map.Entry<String, List<String>> entry : patternSpec.getMethods().entrySet()) {
-                    String cls = entry.getKey();
-                    for (String method : entry.getValue()) {
-                        filter.includeTest(cls, method);
+        Set<Test> cleared = new HashSet<>();
+        for (InternalTaskSpec taskSpec : testExecutionRequest.getTaskSpecs()) {
+            if (taskSpec instanceof InternalTestSpec) {
+                InternalTestSpec testSpec = (InternalTestSpec) taskSpec;
+                Set<Test> tasks = queryTestTasks(taskSpec.getTaskPath());
+                for (Test task : tasks) {
+                    maybeClearFilter(task, cleared); // we clear existing filters defined by the build to ensure that the Tooling API client has full control over what tests to execute
```

Comment:
❓ Do we need to clear filters?

The user might want to experiment with running selected tests in the context of a test task that doesn't include the test by default. I'm not sure though if it's enough justification to clear existing filters.